v5_7_latest command exited with code 127

Server does not start
```
java.lang.RuntimeException: Command exited with error code: '127' and output: ''
[info]   at com.wix.mysql.distribution.setup.ProcessRunner.resolveException(ProcessRunner.java:50)
[info]   at com.wix.mysql.distribution.setup.ProcessRunner.run(ProcessRunner.java:35)
[info]   at com.wix.mysql.distribution.setup.Mysql57Initializer.apply(Mysql57Initializer.java:35)
[info]   at com.wix.mysql.distribution.Setup.apply(Setup.java:26)
[info]   at com.wix.mysql.MysqldExecutable.start(MysqldExecutable.java:42)
[info]   at com.wix.mysql.MysqldExecutable.start(MysqldExecutable.java:21)
[info]   at de.flapdoodle.embed.process.runtime.Executable.start(Executable.java:101)
[info]   at com.wix.mysql.EmbeddedMysql.<init>(EmbeddedMysql.java:47)
[info]   at com.wix.mysql.EmbeddedMysql$Builder.start(EmbeddedMysql.java:155)
```

I have same problem using this library in Spring Boot application v1.5.1.RELEASE

- **os version:** 16.04.1-Ubuntu
- **java version:** 1.8.0_161

| lib version | mysql version | error code |
| - | - | - |
| 3.1.0 | v5_7_19 | 127 |
| 3.1.0 | v5_7_18 | 127 |
| 3.1.0 | v5_6_latest | 141 |
| 3.0.0 | v5_7_19 | 127 |
| 2.2.10 | v5_7_19 | 127 |

**Output for 3.1.0 and v5_7_19 :**
```
    Download Version 5.7.19:Linux:B64 DONE
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz START
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/bin/mysql
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/bin/mysqladmin
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/bin/my_print_defaults
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/bin/mysqld
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/bin/resolveip
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/share/english/errmsg.sys
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/share/fill_help_tables.sql
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/share/mysql_security_commands.sql
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/share/mysql_system_tables.sql
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz extract mysql-5.7.19-linux-glibc2.12-x86_64/share/mysql_system_tables_data.sql
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz nothing left
    Extract /tmp/MySQL-5.7/mysql-5.7.19-linux-glibc2.12-x86_64.tar.gz DONE
   
    .
    .
    .
    
    2018-01-31 16:43:51.514 $ threadId: ERROR  namespace:com.example o.s.b.SpringApplication [ hostname: serviceName: version: transactionId: conversationId: requestTimeStamp:  responseTimeStamp: duration:] Application startup failed
    org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'appEntityManagerFactory' defined in class path resource [com/example/config/AppDatabaseConfig.class]: Unsatisfied dependency expressed through method 'appEntityManagerFactory' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration': Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'embeddedMysqlDb' defined in class path resource [com/example/config/EmbeddedMysqlConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.wix.mysql.EmbeddedMysql]: Factory method 'embeddedMysql' threw exception; nested exception is java.lang.RuntimeException: Command exited with error code: '127' and output: ''
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:467)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
        at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1081)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:856)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:542)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:737)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:370)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:314)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1162)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1151)
        at com.example.Application.main(Application.java:47)
    Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration': Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'embeddedMysqlDb' defined in class path resource [com/example/config/EmbeddedMysqlConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.wix.mysql.EmbeddedMysql]: Factory method 'embeddedMysql' threw exception; nested exception is java.lang.RuntimeException: Command exited with error code: '127' and output: ''
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749)
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:189)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1193)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1095)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:372)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:208)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1138)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:835)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741)
        ... 19 common frames omitted
    Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'embeddedMysqlDb' defined in class path resource [com/example/config/EmbeddedMysqlConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.wix.mysql.EmbeddedMysql]: Factory method 'embeddedMysql' threw exception; nested exception is java.lang.RuntimeException: Command exited with error code: '127' and output: ''
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:296)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:208)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1138)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:835)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741)
        ... 42 common frames omitted
    Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.wix.mysql.EmbeddedMysql]: Factory method 'embeddedMysql' threw exception; nested exception is java.lang.RuntimeException: Command exited with error code: '127' and output: ''
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
        ... 57 common frames omitted
    Caused by: java.lang.RuntimeException: Command exited with error code: '127' and output: ''
        at com.wix.mysql.distribution.setup.ProcessRunner.resolveException(ProcessRunner.java:50)
        at com.wix.mysql.distribution.setup.ProcessRunner.run(ProcessRunner.java:35)
        at com.wix.mysql.distribution.setup.Mysql57Initializer.apply(Mysql57Initializer.java:35)
        at com.wix.mysql.distribution.Setup.apply(Setup.java:26)
        at com.wix.mysql.MysqldExecutable.start(MysqldExecutable.java:42)
        at com.wix.mysql.MysqldExecutable.start(MysqldExecutable.java:21)
        at de.flapdoodle.embed.process.runtime.Executable.start(Executable.java:101)
        at com.wix.mysql.EmbeddedMysql.<init>(EmbeddedMysql.java:46)
        at com.wix.mysql.EmbeddedMysql$Builder.start(EmbeddedMysql.java:160)
        at com.example.config.EmbeddedMysqlConfig.embeddedMysql(EmbeddedMysqlConfig.java:37)
        at com.example.config.EmbeddedMysqlConfig$$EnhancerBySpringCGLIB$$b1f46ab.CGLIB$embeddedMysql$0(<generated>)
        at com.example.config.EmbeddedMysqlConfig$$EnhancerBySpringCGLIB$$b1f46ab$$FastClassBySpringCGLIB$$74be74ed.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:356)
        at com.example.config.EmbeddedMysqlConfig$$EnhancerBySpringCGLIB$$b1f46ab.embeddedMysql(<generated>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
        ... 58 common frames omitted
```


**Usage:**


    @Configuration
    public class EmbeddedMysqlConfig {
    
        public static final String dbName = "embeddedMysqlDb";
    
        @Bean(dbName)
        @Profile("local")
        public EmbeddedMysql embeddedMysql() {
            MysqldConfig config = aMysqldConfig(Version.v5_7_19)
                    .withCharset(Charset.UTF8)
                    .withPort(3306)
                    .build();
            DownloadConfig downloadConfig = aDownloadConfig()
                    .withProxy(aHttpProxy("<redacted>", 8888))
                    .withCacheDir(System.getProperty("java.io.tmpdir"))
                    .build();
            return anEmbeddedMysql(config, downloadConfig)
                    .addSchema("whiteboard-camunda")
                    .addSchema("whiteboard-app")
                    .addSchema("whiteboard-history")
                    .start();
        }
    
        @Bean(dbName)
        @Profile("!local")
        public EmbeddedMysql fakeMysql() {
            return null;
        }
    }

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/distribution/setup/ProcessRunner.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/distribution/setup/ProcessRunner.java
@@ -1,38 +1,41 @@
 package com.wix.mysql.distribution.setup;
 
 import com.wix.mysql.exceptions.MissingDependencyException;
-import com.wix.mysql.io.NotifyingStreamProcessor;
 import com.wix.mysql.io.TimingOutProcessExecutor;
 import de.flapdoodle.embed.process.config.IRuntimeConfig;
 import de.flapdoodle.embed.process.io.IStreamProcessor;
 import de.flapdoodle.embed.process.io.Processors;
+import de.flapdoodle.embed.process.io.ReaderProcessor;
 import de.flapdoodle.embed.process.io.StreamToLineProcessor;
-import org.apache.commons.io.IOUtils;
 
 import java.io.IOException;
 import java.io.InputStreamReader;
-import java.util.concurrent.TimeUnit;
 
 import static java.lang.String.format;
 
 final class ProcessRunner {
 
-    final TimingOutProcessExecutor tope;
+    private final TimingOutProcessExecutor tope;
 
     ProcessRunner(String cmd) {
         this.tope = new TimingOutProcessExecutor(cmd);
     }
 
     void run(Process p, IRuntimeConfig runtimeConfig, long timeoutNanos) throws IOException {
-        IStreamProcessor outputWatch = StreamToLineProcessor.wrap(runtimeConfig.getProcessOutput().getOutput());
+        CollectingAndForwardingStreamProcessor wrapped =
+                new CollectingAndForwardingStreamProcessor(runtimeConfig.getProcessOutput().getOutput());
+        IStreamProcessor loggingWatch = StreamToLineProcessor.wrap(wrapped);
+
         try {
-            Processors.connect(new InputStreamReader(p.getInputStream()), outputWatch);
-            Processors.connect(new InputStreamReader(p.getErrorStream()), outputWatch);
+            ReaderProcessor processorOne = Processors.connect(new InputStreamReader(p.getInputStream()), loggingWatch);
+            ReaderProcessor processorTwo = Processors.connect(new InputStreamReader(p.getErrorStream()), loggingWatch);
 
             int retCode = tope.waitFor(p, timeoutNanos);
 
             if (retCode != 0) {
-                resolveException(retCode, IOUtils.toString(p.getInputStream()) + IOUtils.toString(p.getErrorStream()));
+                processorOne.join(10000);
+                processorTwo.join(10000);
+                resolveException(retCode, wrapped.getOutput());
             }
 
         } catch (InterruptedException e) {
@@ -48,6 +51,28 @@ final class ProcessRunner {
                             "For details see: http://bugs.mysql.com/bug.php?id=60544");
         } else {
             throw new RuntimeException(format("Command exited with error code: '%s' and output: '%s'", retCode, output));
+        }
+    }
+
+    public static class CollectingAndForwardingStreamProcessor implements IStreamProcessor {
+        volatile String output = "";
+        final IStreamProcessor forwardTo;
+
+        CollectingAndForwardingStreamProcessor(IStreamProcessor forwardTo) {
+            this.forwardTo = forwardTo;
+        }
+
+        public void process(String block) {
+            output = output + block;
+            forwardTo.process(block);
+        }
+
+        public void onProcessed() {
+            forwardTo.onProcessed();
+        }
+
+        String getOutput() {
+            return output;
         }
     }
 }

